plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/concurrent/ConcurrentIssueRegistry.java

```diff
@@ -0,0 +1,323 @@
+			if (containerIssues == null) {
+				return false;
+			}
+			persistedIssues.remove(uri);
+			ImmutableSortedSet<LSPIssue> issuesOld = containerIssues.remove(uri);
+			event = issuesOld != null ? eventPersisted(containerHandle, uri, issuesOld, null) : null;
+		}
+		if (event != null) {
+			notifyListeners(event);
+			return true;
+		}
+		return false;
+	}
+
+	public void setIssuesOfDirtyState(URI uri, Iterable<? extends LSPIssue> issues) {
+		ImmutableSortedSet<LSPIssue> issuesNew = ImmutableSortedSet.copyOf(issueComparator, issues);
+		ImmutableSortedSet<LSPIssue> issuesOld;
+		synchronized (this) {
+			issuesOld = dirtyIssues.put(uri, issuesNew);
+		}
+		notifyListeners(eventDirty(uri, issuesOld, issuesNew));
+	}
+
+	public void setIssuesOfPersistedState(String containerHandle, URI uri, Iterable<? extends LSPIssue> issues) {
+		ImmutableSortedSet<LSPIssue> issuesNew = ImmutableSortedSet.copyOf(issueComparator, issues);
+		ImmutableSortedSet<LSPIssue> issuesOld;
+		synchronized (this) {
+			Map<URI, ImmutableSortedSet<LSPIssue>> containerIssues = getContainerIssues(containerHandle, true);
+			persistedIssues.put(uri, issuesNew);
+			issuesOld = containerIssues.put(uri, issuesNew);
+		}
+		notifyListeners(eventPersisted(containerHandle, uri, issuesOld, issuesNew));
+	}
+
+	public synchronized void addIssueOfPersistedState(String containerHandle, URI uri, LSPIssue issue) {
+		ImmutableSortedSet<LSPIssue> issuesOld;
+		ImmutableSortedSet<LSPIssue> issuesNew;
+		synchronized (this) {
+			Map<URI, ImmutableSortedSet<LSPIssue>> containerIssues = getContainerIssues(containerHandle, true);
+			issuesOld = containerIssues.get(uri);
+			issuesNew = ImmutableSortedSet.orderedBy(issueComparator)
+					.addAll(issuesOld != null ? issuesOld : Collections.emptyList())
+					.add(issue).build();
+			persistedIssues.put(uri, issuesNew);
+			containerIssues.put(uri, issuesNew);
+		}
+		notifyListeners(eventPersisted(containerHandle, uri, issuesOld, issuesNew));
+	}
+
+	protected synchronized Map<URI, ImmutableSortedSet<LSPIssue>> getContainerIssues(String containerHandle,
+			boolean createIfNecessary) {
+
+		Map<URI, ImmutableSortedSet<LSPIssue>> containerIssues = container2persistedIssues.get(containerHandle);
+		if (containerIssues == null && createIfNecessary) {
+			containerIssues = new HashMap<>();
+			container2persistedIssues.put(containerHandle, containerIssues);
+		}
+		return containerIssues;
+	}
+
+	public synchronized void addListener(IIssueRegistryListener l) {
+		listeners.add(l);
+	}
+
+	public synchronized void removeListener(IIssueRegistryListener l) {
+		listeners.remove(l);
+	}
+
+	protected void notifyListeners(IssueRegistryChangeEvent event) {
+		notifyListeners(Collections.singletonList(event));
+	}
+
+	protected void notifyListeners(List<IssueRegistryChangeEvent> events) {
+		List<IIssueRegistryListener> _listeners;
+		synchronized (this) {
+			_listeners = new ArrayList<>(listeners);
```

Comment:
listeners could be a CopyOnWriteArrayList which would often avoid unnecessary pessimistic copies. Not sure how many listeners we expect, though.

Reply:
True, or a concurrent list could be used for `listeners`. But when making classes thread safe I usually try to stick to one of two approaches: either make fields thread safe by using concurrent collections, etc. or use synchronize on/in methods. That's why I wanted to stick to using "synchronize" even in the simple case of field `listeners` and even used `this` as argument to synchronize instead of just `listeners`. But this might be overly cautious, idk. What do you think?